### PR TITLE
feat(gemma4): cross the sliding window — front-release with retained global

### LINKED ISSUE #886: gemma4: serving is fail-closed at the sliding window
## Description

The KV serving path refuses any request that reaches the sliding window. Below it, a cache slot and an absolute position are the same number, which is what lets the paged preps write at `start_pos + token` and the attention plan address pages directly. The moment the window moves, those two coordinate systems separate, and nothing in the path expresses the difference — so the path stops rather than silently attending to tokens that have aged out.

Crossing it is also where the local family's completed-step occupancy stops growing: sliding attention only ever reads the last `window` keys, so a request past the window holds pages it can never read again, while the global family legitimately keeps every page it has. A step still peaks at everything it just wrote, and the pool's buffer is allocated up front either way — what changes is how many pages a request holds once the step is done.

## Proposed Plan

- Hold the local family's pages as one reservation each, so the front can be released page by page once a page's last token has aged out of every future window.
- Give the paged preps a `page_origin` alongside the resident page row: the origin is page-aligned, so in-page offsets stay position-invariant and only the row index shifts. RoPE keeps absolute positions.
- Build the local attention plan in cache-relative coordinates, and let `window_left` mask whatever sub-window prefix the first resident page still carries.
- Release after the step's attention, not before: append-then-attend keeps the old window and the new tokens co-resident through the layers.

## Acceptance Criteria

- A request that crosses the window matches the reference distribution at 1023/1024/1025/4096 prompt lengths.
- The same crossing request generates identical tokens with front-release enabled and disabled, while holding strictly fewer pages.
- The local family's completed-step logical resident pages stay bounded once the window engages — a step's own transient peak is separate — while the global family keeps every page.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #886

The local family becomes a sliding state: resident pages held as single-page reservations, released from the front once `(page + 1) * 16 + window <= frontier` — after the step's attention, since append-then-attend keeps the old window and the new tokens co-resident through the layers. What stops growing is the *logical* page count a request holds once a step completes; the step itself still peaks at everything it wrote, and the pool's device buffer is allocated up front either way, so a release returns pages to the pool rather than to the driver. The paged preps take the resident row plus a `page_origin`: the origin is page-aligned, so in-page offsets stay position-invariant and only the row index shifts. RoPE keeps absolute positions; the attention plan lives in cache-relative coordinates, where `window_left` masks whatever sub-window prefix the first page still carries — a non-aligned resident start loses nothing.

## Test Env

Single GPU (sm_89, x86_64), CUDA 12.9, pinned 12B checkpoint, `--test-threads=1`

## Verification

- clippy `-D warnings` across gemma4/core/kernels targets — clean.
- **Window crossing, distribution-level**: greedy chains are not reference material at this depth — measured per-step top1-top2 margins collapse to 0.0-1.3 past ~1000 tokens for every corpus tried, and HF's own sdpa/eager continue the same prompt in different directions (their teacher-forced top-1 self-agreement is 6/9 per case). The fixture records teacher-forced top-64 logprobs under BOTH backends at 1023/1024/1025/4096 prompt lengths (+8 forced steps each); the gate asserts our path within 2x the measured cross-backend gap. Results: max |dlogprob| 0.97 / 2.62 / 1.35 / 4.27 against tolerances 2.92 / 4.61 / 4.99 / 7.65 — the first and third sit BELOW the backends' own gap — with top-1 agreement 7/8/8/6 of 9, at or above HF's self-agreement.
- **Chunked crossing**: the 4096 case re-run in window-sized chunks, which is the only shape that reaches a multi-token step with the resident row already shifted — max |dlogprob| 2.55 against the same 7.65 tolerance, top-1 8/9, and the test asserts such a step actually occurred rather than trusting the arithmetic.
- **Release preserves the generated tokens while reducing the logical footprint**: the same crossing request with front-release disabled generates identical tokens while holding strictly more pages (65 vs 66 at the test length). The A/B compares tokens, not logits.
- **Footprint is asserted exactly, not loosely**: each case asserts `ceil(kv_len / 16) - floor((kv_len - window) / 16)` resident pages, so releasing one page late fails the gate.
- **Footprint curves**: the local family's *logical* resident pages, counted once a step completes, stop growing while global retains everything. For the 4096 case: the prefill step peaks at 256 local pages and ends at local 64 against global 256; after the eight teacher-forced tokens (kv_len 4104) it is local 65 against global 257. Two boundaries on that claim: the pool's device buffer is allocated up front, so releasing pages returns them to the pool's free list rather than lowering process HBM; and a single 4096-token prefill still holds all 256 local pages at once until that step's attention is done, since the front is only released afterwards.
- KV12 gates re-run green over the sliding refactor (regression): 25 library tests plus the six checkpoint gates from #885, unchanged.
- **What each case actually exercises**: front-release only triggers once `(origin + 1) * 16 + window <= frontier`, i.e. at frontier 1040, so w1023/w1024/w1025 (1031/1032/1033 tokens after their forced steps) gate the *mask* boundary — `window_left` starting to hide the oldest keys — while **w4096 is the case that releases**, and it is where the footprint claim comes from (local 65 pages against global 257). The eviction A/B runs at the same shorter length, so its margin is one page (65 vs 66); the wide margin lives in w4096.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### CODE REVIEW

File: tools/accuracy/dump_gemma4_window_golden.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+"""Dump the Gemma 4 window-crossing logprob reference.
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply tools/accuracy/dump_gemma4_window_golden.py:1]: Sure

**Comment by FeathBow** [comment]
@codex review please

**Comment by FeathBow** [comment]
@codex review please

File: pegainfer-gemma4/src/serve_oracle.rs [hunk 2]
```diff
@@ -40,16 +42,282 @@ fn fixture_manifest(bytes: &[u8], key: &str) -> serde_json::Value {
...
+            "{label}: resident pages after {released} released"
+        );
+        assert_eq!(
+            run.global_pages,
+            run.kv_len.div_ceil(page),
+            "{label}: the global family must keep every page"
+        );
+        if max_abs > tolerance {
+            over.push(format!("{label} ({max_abs} > {tolerance})"));
+        }
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/serve_oracle.rs:290]: done
